bots/kablamo.rb

```diff
@@ -0,0 +1,97 @@
+class Kablamo
+
+  def initialize()
+  end
+
+  def display_name()
+    return "Kablamo"
+  end
+
+  def act(game_state, bot_state)
+
+    bot_x = bot_state.position.x
+    bot_y = bot_state.position.y
+
+    enemy_x = game_state.enemies[0].position.x
+    enemy_y = game_state.enemies[0].position.y
+
+    # charge battery
+    if bot_state.available_energy < 5
+       return :charge_battery
+    end
+
+    # Doge the lasers
+    if game_state.enemies[0].available_energy < 5
+      if bot_x - 1 == enemy_x || bot_x + 1 == enemy_x
+        if bot_y < enemy_y
+          if game_state.enemies[0].facing == :north
+            if bot_x > 3 && bot_x < game_state.arena_size.width - 4 && bot_y > 3 && game_state.arena_size.height - 4
+              return :lunge
+            else
+              return :reverse
+            end
+          end
+        else
+          if bot_y > enemy_y
+            if game_state.enemies[0].facing == :south
+              if bot_x > 3 && bot_x < game_state.arena_size.width - 4 && bot_y > 3 && game_state.arena_size.height - 4
+                return :lunge
+              else
+                return :reverse
+              end
+            end
+          end
+        end
+      else
+        if bot_y - 1 == enemy_y || bot_y + 1 == enemy_y
+          if bot_x < enemy_x
+            if game_state.enemies[0].facing == :west
+              if bot_x > 3 && bot_x < game_state.arena_size.width - 4 && bot_y > 3 && game_state.arena_size.height - 4
+                return :lunge
+              else
+                return :reverse
+              end
+            end
+          else
+            if bot_x > enemy_x
+              if game_state.enemies[0].facing == :east
+                if bot_x > 3 && bot_x < game_state.arena_size.width - 4 && bot_y > 3 && game_state.arena_size.height - 4
+                  return :lunge
+                else
+                  return :reverse
+                end
+              end
+            end
+          end
+        end
+      end
+    end
+
+   # Track enemy on the x axis
+    if bot_x < enemy_x && bot_state.facing != :east
+      return :face_east
+    elsif bot_x > enemy_x && bot_state.facing != :west
+      return :face_west
+    end
+
+    # Attack if enemy is already at the same y as me
```

Comment:
@BradyDiffner: Theoretically, won't the bot fire even if it's not facing the enemy? If the bot is facing south and the enemy bot comes from the north _even in the same y axis_, I believe this bot would still shoot. Does that make sense?This might be a feature not a bug, but it seems like a bug.

Reply:
I got confused by this too at first, but actually what you pointed out would only be the case if it was "if bot_x == enemy_x" instead of "if bot_y == enemy_y". If the bots are on the same y coordinate, one cannot come from above the other. I know that "if bot_y == enemy_y return :fire_laser” would be problematic by itself, but the code that gets executed before it ensures that it is facing the fight way before it checks to see if the enemy is on the same y axis.
Hope this clears things up.

Reply:
> If the bots are on the same y coordinate, one cannot come from above the other

Theoretically though, couldn't a bot come from below while your bot is facing west, east, or north? Then the bot would be in the same y axis but your bot is not facing south towards the bot. Would your bot still fire because it is in the same y axis? Or is that what the code on 87-94 takes care of?

> but the code that gets executed before it ensures that it is facing the fight way before it checks to see if the enemy is on the same y axis.

All the code I see before it is for the same x axis and for lunges.

All this is actually probably really wrong, because I'm sure you've tested it, but I'm still vaguely curious.

Reply:
I think whats confusing you is the transition from lines 70-75 to 77-80. While the first deals with the x axis and the later with y, they are both closely connected. Remember that if the both bots are on the same y, they will be shooting in an x direction. 

Here's a more detailed description for each of the four chunks of code that deal with tracking:

70-75
Look towards enemy on the x axis.

77-80
attack if enemy is in my line of fire _across_ the x axis (which would mean they have the same y as me).

82-85
If this is not the case, advance until we have the same x.(So that I can fire across the y axis).

87-94
face the correct y direction and fire.

So in the case a bot comes up beneath me, 70-5 gets ignored because our x is the same,
77-80 is ignored because that deals with y, not x, 82-85 again gets ignored because our x is the same, and 87-94 would make the bot look down towards the enemy and fire, which is the intended result.

Hopefully this makes more sense?

Reply:
Aha. @BradyDiffner: Thank you, that clears things up. 👍